sklearn/utils/testing.py

```diff
@@ -398,6 +405,79 @@ def assert_raise_message(exception, message, function, *args, **kwargs):
         assert_in(message, error_message)
 
 
+def assert_not_raises(given_exception, function, *args, **kwargs):
+    """Helper function to make sure that the given error was not raised"""
+
+    try:
+        function(*args, **kwargs)
+    except Exception as e:
+        if isinstance(e, given_exception):
+            raise AssertionError("%r was raised" % e)
+
+
+def _assert_same_model_method(method, X, estimator1, estimator2, msg=None):
+    if hasattr(estimator1, method):
+        m = '%r\n\nhas %s, but\n\n%r\n\ndoes not' % (estimator1, 
+                                                     method,
+                                                     estimator2)
+        assert hasattr(estimator2, method), m        
+        
+        # assert_array_almost_equal does not accept msg argument.
+        # So we manually raise an AssertionError with msg
+        try:
+            assert_array_almost_equal(getattr(estimator1, method)(X), 
+                                      getattr(estimator2, method)(X), 2,)
+        except AssertionError as e:
+            if msg is None:
+                msg = ("Models are not equal. \n\n%s method returned"
+                       " different results for :\n\n%r \n\nand :\n\n%r"
+                       "\n\nException from testing the result was :\n\n"
+                       "%r" % (method, estimator1, estimator2, e))
+
+            raise AssertionError(msg)
+
+
+def _assert_not_same_model_method(method, X, estimator1, estimator2, msg=None):
+    if hasattr(estimator1, method):
+        m = '{!r}\n\nhas {!s}, but\n\n{!r}\n\ndoes not'.format(estimator1, 
+                                                               method,
+                                                               estimator2)
+        # Even though we check for different models,
+        # AssertionError is raised if method present only in estimator1
+        assert hasattr(estimator2, method), m        
+       
+        same_model = False
+
+        # Check if the method(X) returns the same for both models.
+        try:
+            assert_array_almost_equal(getattr(estimator1, method)(X), 
+                                      getattr(estimator2, method)(X), 2,)
+            same_model = True
+        except AssertionError:
+            # If the assertion of equality in results fails, do not modify
+            # same_model
+            pass
+
+        if same_model:
+            if msg is None:
+                msg = ("Models are equal. \n\n%s method returned"
+                       " same results for :\n\n%r \n\nand :\n\n%r" % 
+                       (method, estimator1, estimator2))
+            raise AssertionError(msg)
```

Comment:
@jnothman [ Sorry for commenting again - had force pushed over this ]
Does this look better now?